cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,853 @@
+    size_t num_edges_with_triangles{0};
+
+    while (true) {
+    // Remove edges that have a triangle count of zero. Those should not be accounted
+    // for during the unroling phase.
+    auto edges_with_triangle_last =
+      thrust::stable_partition(handle.get_thrust_policy(),
+                               edge_triangle_count_pair_first,
+                               edge_triangle_count_pair_first + num_triangles.size(),
+                               [k] __device__(auto e) {
+                                 auto num_triangles = thrust::get<1>(e);
+                                 return num_triangles > 0;
+                               });
+    
+    num_edges_with_triangles = static_cast<size_t>(
+      thrust::distance(edge_triangle_count_pair_first, edges_with_triangle_last));
+    
+
+    
+    edgelist_srcs.resize(num_edges_with_triangles, handle.get_stream());
+    edgelist_dsts.resize(num_edges_with_triangles, handle.get_stream());
+    num_triangles.resize(num_edges_with_triangles, handle.get_stream());
+
+    // 'invalid_edge_first' marks the beginning of the edges to be removed
+    auto invalid_edge_first =
+      thrust::stable_partition(handle.get_thrust_policy(),
+                               edge_triangle_count_pair_first,
+                               edge_triangle_count_pair_first + num_triangles.size(),
+                               [k] __device__(auto e) {
+                                 auto num_triangles = thrust::get<1>(e);
+                                 return num_triangles >= k - 2;
+                               });
+    
+    num_invalid_edges = static_cast<size_t>(
+      thrust::distance(invalid_edge_first, edge_triangle_count_pair_first + edgelist_srcs.size()));
+
+    if (num_invalid_edges == 0){
+        break;
+    }
+    
+    auto num_valid_edges = edgelist_srcs.size() - num_invalid_edges;
+
+    // copy invalid edges
+    auto invalid_edges_buffer = allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(
+      num_invalid_edges, handle.get_stream());
+    
+
+
+    
+    thrust::copy(handle.get_thrust_policy(),
+                 thrust::make_zip_iterator(edgelist_srcs.begin() + num_valid_edges, edgelist_dsts.begin() + num_valid_edges),
+                 thrust::make_zip_iterator(edgelist_srcs.begin() + edgelist_srcs.size(), edgelist_dsts.begin() + edgelist_srcs.size()),
+                 get_dataframe_buffer_begin(invalid_edges_buffer));
+    
+    // resize the 'edgelist_srcs' and 'edgelist_dst'
+    edgelist_srcs.resize(num_valid_edges, handle.get_stream());
+    edgelist_dsts.resize(num_valid_edges, handle.get_stream());
+    num_triangles.resize(num_valid_edges, handle.get_stream());
+    
+      
+    // sort back the edge as those are needed later when running a binary tree
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        edge_first,
+                        edge_first + edgelist_srcs.size(),
+                        num_triangles.begin());
```

Comment:
Right, because of the stable partition we no longer need to sort the edge list again because the former will keep it sorted even when pushing invalid edges to the second partition.